[Merged by Bors] - feat(algebra): define "Euclidean" absolute values

Vierkantor opened the pull request:
We say an absolute value `abv : absolute_value R S` is Euclidean if agrees with the Euclidean domain structure on `R`, namely `abv x < abv y ↔ euclidean_domain.r x y`.

Examples include `abs : ℤ → ℤ` and `λ (p : polynomial Fq), (q ^ p.degree : ℤ)`, where `Fq` is a finite field with `q` elements. (These two correspond to the number field and function field case respectively, in our proof that the class number of a global field is finite.)

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
feat(algebra): define "Euclidean" absolute values

A euclidean absolute value `abv : euclidean_absolute_value R S` is an absolute value that agrees with the Euclidean domain structure on `R`, namely `abv x < abv y ↔ euclidean_domain.r x y`.

Examples include `abs : ℤ → ℤ` and `λ (p : polynomial Fq), (q ^ p.degree : ℤ)`, where `Fq` is a finite field with `q` elements.

Vierkantor commented on src/algebra/euclidean_absolute_value.lean:
```diff
@@ -0,0 +1,144 @@
+
+## Main definitions
+
+ * `euclidean_absolute_value R S` is the type of absolute values on `R` mapping to `S`.
```
```suggestion
 * `euclidean_absolute_value R S` is the type of absolute values on `R` mapping to `S`
    that preserve the order on `R` arising from the Euclidean domain structure.
```

Vierkantor committed:
Fix docstring

jcommelin commented on src/algebra/euclidean_absolute_value.lean:
```diff
@@ -0,0 +1,145 @@
+    that preserve the order on `R` arising from the Euclidean domain structure.
+ * `euclidean_absolute_value.abs` is the "standard" absolute value on `ℤ`,
+   mapping negative `x` to `-x`.
+ * `euclidean_absolute_value.card_pow_degree` is an Euclidean absolute value on
```
```suggestion
 * `euclidean_absolute_value.card_pow_degree` is a Euclidean absolute value on
```
I think...

jcommelin commented on src/algebra/euclidean_absolute_value.lean:
```diff
@@ -0,0 +1,145 @@
+`euclidean_domain` structure on `R`, namely `abv` is strictly monotone with respect to the well
+founded relation `≺` on `R`. -/
+structure euclidean_absolute_value (R S : Type*) [euclidean_domain R] [ordered_semiring S]
+  extends absolute_value R S :=
```
Should this `extend` or should it be a predicate on `absolute_value`s?

Vierkantor replied:
It does seem easier to write out the API, at least. The real test will be to implement `absolute_value.is_admissible` in the upcoming PR :)

Vierkantor committed:
Make `euclidean_absolute_value` a predicate on absolute values

I still need the predicate to be a structure since it will be extended
to `is_admissible` in the finiteness proof of the class number.

Vierkantor committed:
Add module doc

robertylewis commented on src/algebra/euclidean_absolute_value.lean:
```diff
@@ -0,0 +1,63 @@
+
+end is_euclidean
+
+section int
+
+open int
```
```suggestion
namespace int
```
?

jcommelin reviewed (approved):
Thanks :tada:

bors merge
